bundles/org.openhab.core.io.rest.update/src/main/java/org/openhab/core/io/rest/update/internal/factory/UpdaterFactory.java

```diff
@@ -0,0 +1,114 @@
+/**
+ * Copyright (c) 2010-2021 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.core.io.rest.update.internal.factory;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Properties;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.io.rest.update.internal.enums.OperatingSystem;
+import org.openhab.core.io.rest.update.internal.enums.PackageManager;
+import org.openhab.core.io.rest.update.internal.updaters.BaseUpdater;
+import org.openhab.core.io.rest.update.internal.updaters.DebianUpdater;
+import org.openhab.core.io.rest.update.internal.updaters.MacUpdater;
+import org.openhab.core.io.rest.update.internal.updaters.PacManUpdater;
+import org.openhab.core.io.rest.update.internal.updaters.PortageUpdater;
+import org.openhab.core.io.rest.update.internal.updaters.WindowsUpdater;
+import org.openhab.core.io.rest.update.internal.updaters.YumUpdater;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link UpdaterFactory} creates a xxxUpdater class based on the Operating System (and its respective package
+ * manager if the Operating System is Linux).
+ *
+ * @author AndrewFG - initial contribution
+ */
+@NonNullByDefault
+public class UpdaterFactory {
+    // logger must be static because the class needs to log within static methods
```

Comment:
I think it would be overkill for the factory to be an Osgi service. The REST resource is a component with precisely one instance, which on creation makes precisely one call to the factory, to create precisely one instance of an xxxUpdater class. So if you don’t want a static factory, I would simply change replace ‘UpdaterFactory.newUpdater()’ with ‘new UpdaterFactory().newUpdater’ in the REST resource constructor. (See also your other comment about constructor injection..)

Reply:
Instantiate both the factory and the respective update class in one line in the REST resource's constructor.